src/oge/helpers.py

```diff
@@ -287,6 +290,52 @@ def create_plant_ba_table(year: int) -> pd.DataFrame:
     return plant_ba
 
 
+def add_plant_operating_and_retirement_dates(df: pd.DataFrame) -> pd.DataFrame:
+    """Adds the operating and retirement dates of a plant to input data frame. The
+    operating date of a plant is taken as the earliest date among all generators'
+    operating date over all report dates. Likewise, the retirement date of a plant is
+    taken as the latest date among all generators' retirement date over all report
+    dates.
+
+    Note that the operating date is the date the generator began commercial operation.
+    The retirement date is the date of the scheduled or effected retirement of the
+    generator.
+
+    Args:
+        df (pd.DataFrame): table with a 'plant_id_eia' column.
+
+    Returns:
+        pd.DataFrame: original data frame with additional 'operating_date' and
+            'retirement_date' column.
+    """
+    generators_dates = load_data.load_pudl_table(
+        "denorm_generators_eia",
+        columns=[
+            "plant_id_eia",
+            "generator_id",
+            "report_date",
+            "generator_operating_date",
+            "generator_retirement_date",
+        ],
+    )
+    plants_dates = (
+        generators_dates.groupby("plant_id_eia")[
+            ["generator_operating_date", "generator_retirement_date"]
+        ]
+        .agg({"generator_operating_date": "min", "generator_retirement_date": "max"})
```

Comment:
While this will work for the operating date, this will not work for the retirement date. For example, what if a plant has 10 generators and only one retires? This would currently say the entire plant is retired.

For the retirement date, one way to do this would be to check for plants where there are no NA retirement dates across all generators, and then take the max of that.

Looking at the sample outputs you posted, it currently shows that plant 3 "Barry" retired in 2015, but this plant is still operational

Reply:
One easy way to do this would be to load the "operational_status' column and just identify where all generators are retired as of the latest_validated_year

Reply:
Indeed

Reply:
Done